### PR TITLE
Remote Attestation Sample Demonstrating Consumption of New Evidence API

### PR DESCRIPTION
**Pull request by kevinxu118**

### CODE REVIEW

**Comment by yentsanglee** [comment]
Can you start a new PR that only include changes in remote_attestation sample?

File: samples/local_attestation/common/attestation.cpp [hunk 1]
```diff
@@ -12,154 +19,159 @@ Attestation::Attestation(Crypto* crypto, uint8_t* enclave_mrsigner)
...
+    // Initialize attester and use the SGX plugin.
+    attester_result = oe_attester_initialize();
+    if (attester_result != OE_OK)
+    {
+        TRACE_ENCLAVE("oe_attester_initialize failed.");
+        goto exit;
+    }
+
+    // Generate evidence based on the format selected by the attester.
+    result = oe_get_evidence(&sgx_local_uuid, NULL, NULL, 0, target_info_buffer, target_info_size, evidence_buf, local_evidence_buf_size, NULL, 0);
```
Comment (qiucwang) [inline samples/local_attestation/common/attestation.cpp:52]: Have you tried to reformat the code? This line seems to be long.
`./scripts/format-code`

File: samples/local_attestation/common/attestation.cpp [hunk 2]
```diff
@@ -12,154 +19,159 @@ Attestation::Attestation(Crypto* crypto, uint8_t* enclave_mrsigner)
...
-    if (memcmp(parsed_report.identity.signer_id, m_enclave_mrsigner, 32) != 0)
+    // Iterate through list of claims.
+    for (size_t i = 0; i < claims_length; i++) 
     {
-        TRACE_ENCLAVE("identity.signer_id checking failed.");
-        TRACE_ENCLAVE(
-            "identity.signer_id %s", parsed_report.identity.signer_id);
-
-        for (int i = 0; i < 32; i++)
+        if (strcmp(claims[i].name, OE_CLAIM_SIGNER_ID) == 0)
```
Comment (qiucwang) [inline samples/local_attestation/common/attestation.cpp:119]: I think the claim check implemented [here](https://github.com/openenclave/openenclave/blob/master/tests/attestation_plugin/plugin/tests.c#L361-L372) might be a better choice. You can abstract a function like `find_claim()` to find the claim first then do the check. It's cleaner and more efficient.
Current approach iterates the claim list once but need 3 `strcmp`s for each claim while `find_claim()` perform 3 times of list iteration but will early stop in each iteration.

File: samples/remote_attestation/common/attestation.cpp [hunk 3]
```diff
@@ -12,153 +18,148 @@ Attestation::Attestation(Crypto* crypto, uint8_t* enclave_mrsigner)
...
-    if (memcmp(parsed_report.identity.signer_id, m_enclave_mrsigner, 32) != 0)
+    // Iterate through list of claims.
+    for (size_t i = 0; i < claims_length; i++) 
     {
-        TRACE_ENCLAVE("identity.signer_id checking failed.");
-        TRACE_ENCLAVE(
-            "identity.signer_id %s", parsed_report.identity.signer_id);
-
-        for (int i = 0; i < 32; i++)
+        if (strcmp(claims[i].name, OE_CLAIM_SIGNER_ID) == 0)
```
Comment (qiucwang) [inline samples/remote_attestation/common/attestation.cpp:113]: Same comment as above

**Comment by yufjiang** [comment]
> Can you start a new PR that only includes changes in remote_attestation sample?

Helped Kevin clean up the PR and now it only has changes on remote attestation sample

File: samples/local_attestation/common/attestation.cpp (hunk 1 above)
Comment (kevinxu118) [reply samples/local_attestation/common/attestation.cpp]: Fixed.

File: samples/local_attestation/common/attestation.cpp (hunk 2 above)
Comment (kevinxu118) [reply samples/local_attestation/common/attestation.cpp]: I agree with that approach. I have updated the code to reflect that.

File: samples/remote_attestation/common/attestation.cpp [hunk 4]
```diff
@@ -2,163 +2,195 @@
...
+    if (result != OE_OK)
     {
-        TRACE_ENCLAVE("identity.signer_id checking failed.");
         TRACE_ENCLAVE(
-            "identity.signer_id %s", parsed_report.identity.signer_id);
+            "oe_verify_evidence failed (%s).\n", oe_result_str(result));
+        goto exit;
+    }
+
+    uint8_t* value;
```
Comment (yentsanglee) [inline samples/remote_attestation/common/attestation.cpp:147]: The convention is to declare all local variables at the beginning of the function.

File: samples/remote_attestation/common/attestation.cpp [hunk 5]
```diff
@@ -2,163 +2,195 @@
...
 
-        for (int i = 0; i < 32; i++)
+    // 1) Validate the signer id.
+    value = (uint8_t*)_find_claim(claims, claims_length, OE_CLAIM_SIGNER_ID);
```
Comment (yentsanglee) [inline samples/remote_attestation/common/attestation.cpp:150]: I'd recommend for _find_claim() to return oe_claim_t* so you know the size of the claim value. Here if the value size is smaller than 32, your comparison at line 152 is questionable.

File: samples/remote_attestation/common/attestation.cpp [hunk 6]
```diff
@@ -2,163 +2,195 @@
...
 
-    // Check the enclave's product id and security version
-    // See enc.conf for values specified when signing the enclave.
-    if (parsed_report.identity.product_id[0] != 1)
-    {
-        TRACE_ENCLAVE("identity.product_id checking failed.");
-        goto exit;
-    }
+    // 2) Check the enclave's product id.
+    value = (uint8_t*)_find_claim(claims, claims_length, OE_CLAIM_PRODUCT_ID);
```
Comment (yentsanglee) [inline samples/remote_attestation/common/attestation.cpp:176]: value can be null here. As commented above, I'd prefer _find_claim to return oe_claim_t*. Same below.

File: samples/remote_attestation/common/attestation.cpp (hunk 4 above)
Comment (kevinxu118) [reply samples/remote_attestation/common/attestation.cpp]: Fixed.

File: samples/remote_attestation/common/attestation.cpp (hunk 5 above)
Comment (kevinxu118) [reply samples/remote_attestation/common/attestation.cpp]: Fixed.

File: samples/remote_attestation/common/attestation.cpp (hunk 6 above)
Comment (kevinxu118) [reply samples/remote_attestation/common/attestation.cpp]: Fixed, same as above.

**Comment by yufjiang** [comment]
bors try

**Comment by yufjiang** [comment]
bors try

**Comment by yentsanglee** [comment]
bors retry

**Pull request merged**